[Refactor] replace std::optional<bool> in InlineFormattingUtils::computedTextIndent with an enum.

yulunwu2 opened the pull request:
#### 58e7b3bcdb877cd51b4f2f17efeb9cecdd4346e4
<pre>
[Refactor] replace std::optional&lt;bool&gt; in InlineFormattingUtils::computedTextIndent with an enum.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291012">https://bugs.webkit.org/show_bug.cgi?id=291012</a>
&lt;<a href="https://rdar.apple.com/148533334">rdar://148533334</a>&gt;

Reviewed by Alan Baradlay.

This PR replaces a std::optional&lt;bool&gt; with an explicit enum class. This makes the way we track
how we laid out the previous line of text much clearer.

Canonical link: <a href="https://commits.webkit.org/293287@main">https://commits.webkit.org/293287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de83622e0923fddd5b0a5094d629ebf2d458963f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8087 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103345 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48757 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100273 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26310 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74764 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31937 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101232 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13835 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88726 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6676 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48199 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83505 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105721 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25314 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18515 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83838 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25687 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84920 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21063 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27856 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5537 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18954 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30447 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25093 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28409 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26668 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
```diff
@@ -154,7 +154,7 @@ InlineLayoutUnit InlineFormattingUtils::computedTextIndent(IsIntrinsicWidthMode
     // [Integration] root()->parent() would normally produce a valid layout box.
     bool shouldIndent = false;
-    if (!previousLineEndsWithLineBreak) {
+    if (previousLineState != PreviousLineState::EndsWithLineBreak) {
```
isn't this NoPreviousLine?

alanbaradlay replied:
but either way, I think having optional to signal that there's no previous line is better than PreviousLineState saying there's no PreviousLineState.

yulunwu2 replied:
Ah, DoesNotEndWithLineBreak should not go down this code path, my mistake. Thanks for catching this!

yulunwu2 committed:
[Refactor] replace std::optional<bool> in InlineFormattingUtils::computedTextIndent with an enum.
https://bugs.webkit.org/show_bug.cgi?id=291012
<rdar://148533334>

Reviewed by Alan Baradlay.

This PR replaces a std::optional<bool> with an explicit enum class. This makes the way we track
how we laid out the previous line of text much clearer.

Canonical link: https://commits.webkit.org/293287@main
